openBalena 2024

ab77 opened the pull request:
take two..

Documentation:
* https://ab77-open-balena.open-balena.pages.dev/
* https://ab77-open-balena.open-balena.pages.dev/getting-started/

balenaHub: https://hub.balena.io/apps/1910724/open-balena

Requires:
* https://github.com/balena-io/open-balena-api/pull/1618
* https://github.com/balena-io/environment-production/pull/2278
* https://github.com/balena-io-modules/balena-compose-parse/pull/63

ab77 commented:
@dfunckt please feel free to give this a go, I linked the very rough getting started guide on balena in the PR description.

The default mDNS configuration will fail at the `balena deploy ...` step, since Docker Desktop on macOS isn't linked with mDNS resolver, so it can't resolve the S3/registry endpoint. This may not be an issue if the balena-cli workstation is running Linux.

Since mDNS configurations are probably only useful for local development, functional public DNS configuration can be built by supplying additional `--env` parameters to the initial `balena push ...` (e.g.):

```
# where DNS_TLD specifies a DNS wildcard entry (A and AAAA) in Gandi or CloudFlare DNS, pointing to the IP of the device hosting openBalena
# both Gandi and CloudFlare DNS are currently supported (pick one)
balena push ${uuid}.local --nocache \
  --env DNS_TLD=balena.acme.com \
  --env mdns:MDNS_TLD='' \
  --env cert-manager:GANDI_API_TOKEN=abcdef1234567890 \
  --env cert-manager:CLOUDFLARE_API_TOKEN=abcdef1234567890 \
  --env cert-manager:ACME_EMAIL=admin@acme.com
```

cc @codewithcheese

dfunckt commented on src/haproxy-sidecar/Dockerfile.template:
```diff
@@ -0,0 +1,7 @@
+FROM docker
```
Why do you need both an `haproxy` sidecar *and* an `haproxy-sidecar` sidecar?

Can their functionality be merged into one repo -- they seem strongly related.

dfunckt commented on src/haproxy-sidecar/balena.sh:
```diff
@@ -0,0 +1,52 @@
+done
+
+while true; do
+    while [ "$(curl --silent --retry 3 --fail \
+      "${BALENA_SUPERVISOR_ADDRESS}/v1/device?apikey=${BALENA_SUPERVISOR_API_KEY}" \
+      -H "Content-Type:application/json" | jq -r '.update_pending')" = 'true' ]; do
+        sleep "$(( (RANDOM % 3) + 3 ))s"
+    done
```
What's the purpose of this?

It'll also only work on balenaOS.

ab77 replied:
Wait fir supervisor to finish syncing target state.

dfunckt commented on src/haproxy/Dockerfile:
```diff
@@ -1,10 +1,3 @@
-VOLUME [ "/certs" ]
-
-RUN apk add --update inotify-tools
+FROM balena/open-balena-haproxy:v4.0.6
```
This is problematic because we're hiding the actual dependency inside a bespoke service. I believe the only reason for this is we need a way to pass configuration. Is there a better way?

There's #127 if you'd rather we tackled this separately.

ab77 replied:
I am not aware of a "better" way, no.

dfunckt replied:
Do we not have a similar use case elsewhere? The VPN? How does it get its config?

ab77 replied:
> This is great progress and a necessary step towards an easily deployable and testable openBalena but the fact it completely disregards support for any OS other than balenaOS makes it a no-go as it is. We really need to work on identifying what is balenaOS-specific and bridge the gap.

Should I close the PR then or will you work on adding other OS support @dfunckt ?

dfunckt commented on src/haproxy/haproxy.cfg:
```diff
@@ -1,134 +1,190 @@
 global
-  tune.ssl.default-dh-param 1024
+	tune.ssl.default-dh-param 1024
```
Can you please switch white-space to 2 spaces to make it easier to see the actual changes?

dfunckt reviewed (commented):
Just briefly went over the PR and left some comments.

ab77 commented:
> This is great progress and a necessary step towards an easily deployable and testable openBalena but the fact it completely disregards support for any OS other than balenaOS makes it a no-go as it is. We really need to work on identifying what is balenaOS-specific and bridge the gap.

Well, for one thing:

```
$ balena push ${uuid}.local
[Info]    Starting build on device 192.168.1.75
Bind mounts are not allowed
```

So that's a bit of a show stopper. Otherwise the second commit in this PR seems to work in compose.

dfunckt commented:
Fix in https://github.com/balena-io-modules/balena-compose-parse/pull/63.

ab77 commented:
@dfunckt this is now ready for a proper review (i.e deployment into both balena and compose environments).

Quickstart addendum WIPs:
* https://github.com/balena-io/open-balena/blob/ab77/open-balena/docs/on-balena.md
* https://github.com/balena-io/open-balena/blob/ab77/open-balena/docs/on-compose.md

... both flows seem to work, as well as the GH Actions tests on ephemeral AWS instances.

The tests are currently implemented using a non-PRed [branch](https://github.com/balena-io-playground/balenaVirt/tree/ab77/open-balena) on balenaVirt. This project would require to be reviewed (incl. nomenclature), moved out of playground, converted into a balenaBlock (e.g. balenablocks/balena-virt) and my branch merged in (with relevant changes). Joseph is the original author of `balenaVirt`.

cc @jakogut @klutchell

klutchell commented:
Sounds good to me!

ab77 commented:
> ... both flows seem to work, as well as the GH Actions [tests](https://github.com/balena-io/open-balena/runs/7460674842?check_suite_focus=true#step:15:530) on ephemeral AWS instances.

needs to be refactored to use Flowzone/custom pipeline (e.g. balena-cloud/Jellyfish), but that should be trivial.

ab77 commented:
Hello @bartversluijs , we are planning to finalise this fairly chunky structural update of openBalena here, and if you are still an active user of this project, I would love your feedback on the [guide](https://ab77-open-balena.open-balena.pages.dev/getting-started/)..

bartversluijs commented:
@ab77 Thanks for the tag! I'll look into it when I have the time. I see in the "Getting started" guide, there are more detailed on making a user and such. Is the basics still the same (like the Docker and HAProxy and such)? I can find out looking at the repo, but I'll look into that later.

I'm more familiair now using it via Kubernetes instead of the direct install on a server with Docker, but I'll definitely check it out!

maurits-inuits commented:
First of all great to see another update for main OpenBalena repository! 
I noticed that automatic certificate renewal only supports the DNS-01 challenge and not the HTTP-01 challenge. Is there a specific reason for this limitation? Seeing as the new cert-manager is based on certbot it should be fairly easy to add, no?

ab77 commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,661 @@
+        device_id="$(balena device "${balena_device_uuid}" | grep ^ID: | cut -c20-)"
+
+        # the actual version deployed depends on the AWS EC2/AMI, defined in AWS_EC2_LAUNCH_TEMPLATE
+        os_version="$(balena os versions ${{ vars.DEVICE_TYPE || 'generic-amd64' }} | head -n 1)"
```
always use latest

ab77 commented on docker-compose.yml:
```diff
@@ -0,0 +1,360 @@
+      REDIS_IS_CLUSTER: 'false'
+      TOKEN_AUTH_JWT_ALGO: ES256
+      TOKENS_CONFIG: API_SERVICE_API_KEY:hex,AUTH_RESINOS_REGISTRY_CODE:hex,COOKIE_SESSION_SECRET:hex,JSON_WEB_TOKEN_SECRET:hex,MIXPANEL_TOKEN:hex,SUPERUSER_PASSWORD:hex,TOKEN_AUTH_BUILDER_TOKEN:hex,VPN_GUEST_API_KEY:hex,VPN_SERVICE_API_KEY:hex,API_VPN_SERVICE_API_KEY:API_SERVICE_API_KEY,REGISTRY2_TOKEN:TOKEN_AUTH_BUILDER_TOKEN
+      TRUST_PROXY: 172.16.0.0/12
```
added for compatibility with newer API versions

ab77 commented:
> First of all great to see another update for main OpenBalena repository! I noticed that automatic certificate renewal only supports the DNS-01 challenge and not the HTTP-01 challenge. Is there a specific reason for this limitation? Seeing as the new cert-manager is based on certbot it should be fairly easy to add, no?

Hello, yes the reason is that we issue a wildcard cert, and these are only supported with `DNS-01` [verification](https://letsencrypt.org/docs/challenge-types/#http-01-challenge). So we chose this approach since it is fairly easy to implement additional DNS providers based on support already provided for Cloudflare and Gandi DNS APIs.

ab77 commented:
> @ab77 Thanks for the tag! I'll look into it when I have the time. I see in the "Getting started" guide, there are more detailed on making a user and such. Is the basics still the same (like the Docker and HAProxy and such)? I can find out looking at the repo, but I'll look into that later.

Thank you, I've basically refactored the guide to make it easier to run cut-paste type deployment and minimise prompts for passwords, etc. Functionally, the user/sudo setup is identical to the previous guide. Other than that, the main changes, apart from the build/test scaffold, are a few extra containers to do automatic SSL/PKI config and wrapping all the scrips into a single `Makefile`.

> I'm more familiair now using it via Kubernetes instead of the direct install on a server with Docker, but I'll definitely check it out!

I am very keen to get your Kubernetes PR in after this, however we are a little "allergic" to Helm over here, so it would be awesome if we could start with just vanilla manifests that can be deployed with kubectl/Flux/Kustomize.

maurits-inuits commented:
Thanks for your reply. I've had the time to go through a large part of the changes and have a few more questions:

1. Am I correct in understanding that the cert-manager now also handles the renewal of the VPN server certificates?

2. I have some security concerns regarding the capabilities being assigned to most of the services, especially when combined with them running as root. I understand that this is due to the use of systemd in the containers, which, according to other GitHub issues, is necessary for your local dev environments. However, would you be open to accepting PR's to create non-systemd images that can be built in parallel through Flowzone? (There's already a `balena/open-balena-base:no-systemd` image available) This would mean that only the VPN would need `NET_ADMIN` capabilities, resulting in a much more secure setup. 

If this is not right place to ask this question please let know.

ab77 commented:
> 1. Am I correct in understanding that the cert-manager now also handles the renewal of the VPN server certificates?

Hello, yes, that is almost correct, except for renewals. There are two components responsible for this, one is [ca-private](https://github.com/balena-io/ca-private), based on Cloudlare's SSL toolkit and the other is [cert-manager](https://github.com/balena-io/cert-manager), which is just a shell script that orchestrates generating self-signed certificates using PKI prepared by ca-private as well as handles ACME exchanges via a few certbot plugins.

The renewal part isn't fully automated (yet). For renewal of VPN certs, as long as the self-signed CA root(s) haven't expired, it's just a matter of deleting the certs in `/certs` on the shared volume and restarting the services. For PKI expiry, there is a provision for rolling new version of the CA via environment variables to ca-private.

> 2. I have some security concerns regarding the capabilities being assigned to most of the services, especially when combined with them running as root. I understand that this is due to the use of systemd in the containers, which, according to other GitHub issues, is necessary for your local dev environments. However, would you be open to accepting PR's to create non-systemd images that can be built in parallel through Flowzone? (There's already a `balena/open-balena-base:no-systemd` image available) This would mean that only the VPN would need `NET_ADMIN` capabilities, resulting in a much more secure setup.

This is a long standing piece of tech. debt. @klutchell and a few other folks here have been talking about replacing `systemd` with more container friendly `s6-init` and/or breaking these `open-balena-base` based monoliths into their own compositions/pods, for a long time. I've personally had a few aborted attempts at splitting `open-balena-vpn` and we've had a few `no-systemd` variants committed to masters without much integration/testing.

We would be more than happy to consider external contributions for this, however, we need to be very clear on which combination of approaches we adopt from the start (i.e. docker-compose, s6 or k8s or all of the above).

otaviojacobi commented on docker-compose.yml:
```diff
@@ -109,6 +109,8 @@ services:
       TRUST_PROXY: 172.16.0.0/12
       VPN_PORT: 443
+      WEBRESOURCES_S3_BUCKET: web-resources
+      WEBRESOURCES_S3_REGION: local
```
@ab77 I think this needs to be us-east-1 as it is the default region minio will use

ab77 replied:
@otaviojacobi how does this region setting affect a local installation of openBalena? Or is there some defaults/hard-coding goingf on which won't work without it being explicitly set to `us-east-1` regardless of actual location?

otaviojacobi replied:
@ab77 This is basically a mandatory parameter to use aws sdk v3 https://github.com/aws/aws-sdk-js-v3/issues/5564#issuecomment-1852890956

Minio will use us-east-1 as default regardless of where it is actually deployed so it can work with aws configurations, therefore we use us-east-1 (as it is the default one used by minio)

ab77 replied:
So if we omit `WEBRESOURCES_S3_REGION`, the default value will be used anyway?

ab77 replied:
this is what I get against the open API:

```
id=$(( RANDOM ))

curl -vi "https://api.$DNS_TLD/resin/release_asset" \
  --header "Authorization: Bearer $(cat <~/.balena/token)" \
  --form 'asset=@"balena.png"' \
  --form "release=\"$id\"" \
  --form 'release asset key="balena.png"'

...
> POST /resin/release_asset HTTP/2
> Host: api.foo.balena-dev.com
> User-Agent: curl/8.5.0
> Accept: */*
> Authorization: Bearer eyJhbG...
> Content-Length: 40608
> Content-Type: multipart/form-data; boundary=------------------------6BlxEVl8hBfWNTBIO9ICrk
> 
* We are completely uploaded and fine
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
< HTTP/2 404 
HTTP/2 404 
< date: Thu, 16 May 2024 20:24:57 GMT
date: Thu, 16 May 2024 20:24:57 GMT
< content-length: 0
content-length: 0
```

robbe-haesendonck commented:
> This is a long standing piece of tech. debt. @klutchell and a few other folks here have been talking about replacing `systemd` with more container friendly `s6-init` and/or breaking these `open-balena-base` based monoliths into their own compositions/pods, for a long time. I've personally had a few aborted attempts at splitting `open-balena-vpn` and we've had a few `no-systemd` variants committed to masters without much integration/testing.

Hello!
I've worked on this before and even had some PR's lined up.
I do still have a question about why systemd (or any init system) is actually necessary, is there any possibility to get rid of it altogether?

I'm certainly willing to pick this up again!

klutchell commented:
@robbe-haesendonck systemd was in use before we split our backend into microservices, and is only used for legacy reasons.

We've determined that indeed some services won't need an initsystem at all and some would benefit from something simple like s6-overlay.

It's just a matter prioritizing the effort with other ongoing work.

dfunckt commented:
We also need systemd because we've been using mdns for local development (and maybe other use cases?) and it pulls in systemd as a dependency and there was is (viable) alternative. This was the state when I last looked into replacing systemd with s6, which is quite a while back (couple of years?) and we may have progressed since.

ab77 commented on docker-compose.yml:
```diff
@@ -0,0 +1,361 @@
+    image: redis:alpine
+    cap_add:
+      - SYS_RESOURCE
+      - SYS_ADMIN
```
> Given the recent changes to the Redis license, I was wondering if there are any concerns about potential legal issues for companies using OpenBalena. Has the team considered any alternatives to Redis in case the license change becomes problematic in the future? I assume you use Redis for Balena Cloud as well.

We are currently considering the implication of this change.

> Also, I'm wondering why a Redis container needs these additional capabilities unless it has something to do with your internal dev environment?

Thank you, removed.

klutchell commented:
I've never seen policy-bot invalidate on push like that. I always thought it was a setting that was only enforced for one approval rule but other approval rules could still pass.

I wonder if it's related to our permissions on this repo?

klutchell commented:
LGTM

klutchell commented:
> I've never seen policy-bot invalidate on push like that. I always thought it was a setting that was only enforced for one approval rule but other approval rules could still pass.
> 
> I wonder if it's related to our permissions on this repo?

Oh, it's because I'm an author. I cannot approve.

ab77 commented:
> Oh, it's because I'm an author. I cannot approve.

We both are now :)

thgreasi commented on docker-compose.yml:
```diff
@@ -0,0 +1,361 @@
+version: '2.4'
+
+volumes:
+  builder-certs-ca: {}
+  builder-certs-client: {}
+  builder-data: {}
```
oB doesn't have a builder so I guess we don't need these volumes, right?

ab77 replied:
Yes, it has a pseudo-builder service (end of composition) for the e2e tests to run.

thgreasi commented on src/tag-sidecar/balena.sh:
```diff
@@ -0,0 +1,44 @@
+			tag_key="$(echo "${kv}" | awk -F';' '{print $1}')"
+			value="$(echo "${kv}" | awk -F';' '{print $2}')"
+
+			curl_with_opts "${BALENA_API_URL}/v6/device_tag" \
```
Just a heads up that this would only initially set the tags and will not be able to further update them again.
We would need to have both a POST & a PATCH in case that we care to keep updating them
See: https://github.com/balena-io-examples/device-tags/blob/master/device_tags/curl_tags_api/set-device-tag.sh

ab77 replied:
I think this could be an improvement at a later stage - we have the same logic in e2e and balena-cloud repos that will need to be made idempotent.

thgreasi commented on docker-compose.yml:
```diff
@@ -0,0 +1,361 @@
+      VPN_PORT: 443
+      WEBRESOURCES_S3_BUCKET: web-resources
+      WEBRESOURCES_S3_REGION: "us-east-1"  # this is required for minio
+      HIDE_UNVERSIONED_ENDPOINT: "false" # unstable/unsupported functionality
```
I might got confused yesterday while discussing this.
What I was trying to say is that I would be fine if we commented this whole line out (or completely remove it)
```suggestion
      # HIDE_UNVERSIONED_ENDPOINT: "false" # uncomment to make the unstable/unsupported /resin endpoint available
```
Having it enabled by default is something I think we should avoid, b/c realistically I don't expect anyone to see this comment before they start using `/resin`. I'm fine w/ having it as I suggested, since at that point users are explicitly made aware that /resin is unstable at the moment that they uncomment this.
If the point is to write tests for release assets, then I highly suggest
* doing it in a separate PR, getting this already big enough PR in sooner
* not doing it until we actually make release assets available on v6/v7, since /resin tests cases might break by definition (exactly how the yocto automation we had that used release assets on /resin did earlier this year).

PS: Just ship it w/o release assets :tada:

thgreasi reviewed (commented):
:heart: The parts that I can understand look fine :+1:

ab77 commented on docker-compose.yml:
```diff
@@ -0,0 +1,361 @@
+      VPN_PORT: 443
+      WEBRESOURCES_S3_BUCKET: web-resources
+      WEBRESOURCES_S3_REGION: "us-east-1"  # this is required for minio
+      HIDE_UNVERSIONED_ENDPOINT: "false"
```
We are still testing this in the GH workflow, but I've removed `HIDE_UNVERSIONED_ENDPOINT` from the composiiton.

ab77 commented:
@thgreasi or @dfunckt when you have a moment, I think we are done here. Please add your final approvals on this rebase/squash and let's get this merged. Then we can resolve/close the open community PRs, if their owners still care and their reality matches ours, and hopefully have this project automatically maintained by Renovate and Co.
